### PR TITLE
Refactors TektonTriggers to use TektonInstallerSet

### PR DESCRIPTION
**Pull request by sm43**
Updates Triggers reconciler to use TektonInstallerSet

This updates the triggers reconciler to use TektonInstallerSet
for installation, this would allow change of TargetNamespace and
deletion of obsolete resources on upgrades.

-----

 Add AllDeploymentsReady Conditions in TektonInstallerSet

If there is any other deployment other than controller and
webhook then it will be missed by installer set for checking the
status so this adds another condition which marks the status of
any other deployment other than controller and webhook.

Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

File: pkg/reconciler/common/deployments_test.go [hunk 1]
```diff
@@ -16,107 +16,93 @@ limitations under the License.
...
+//		ObjectMeta: metav1.ObjectMeta{
+//			Namespace: "test",
+//			Name:      "ready",
+//		},
+//		Status: appsv1.DeploymentStatus{
+//			Conditions: []appsv1.DeploymentCondition{{
+//				Type:   appsv1.DeploymentAvailable,
+//				Status: corev1.ConditionTrue,
+//			}},
+//		},
```
Comment (nikhil-thomas) [inline pkg/reconciler/common/deployments_test.go:30]: could we remove this file completely

File: pkg/reconciler/common/install_test.go [hunk 2]
```diff
@@ -16,181 +16,167 @@ limitations under the License.
...
+//
+//	// Resources in the manifest
+//	targetNamespace := "tekton-pipelines"
+//	wantdeployment := namespacedResource("apps/v1", "Deployment", "test", "test-deployment")
+//	wantrole := namespacedResource("rbac.authorization.k8s.io/v1", "Role", "test", "test-role")
+//	wantroleBinding := namespacedResource("rbac.authorization.k8s.io/v1", "RoleBinding", "test", "test-role-binding")
+//	wantclusterRole := clusterScopedResource("rbac.authorization.k8s.io/v1", "ClusterRole", "test-cluster-role")
+//	wantclusterRoleBinding := clusterScopedResource("rbac.authorization.k8s.io/v1", "ClusterRoleBinding", "test-cluster-role-binding")
+//	deployment := namespacedResource("apps/v1", "Deployment", "test", "test-deployment")
+//	role := namespacedResource("rbac.authorization.k8s.io/v1", "Role", "test", "test-role")
```
Comment (nikhil-thomas) [inline pkg/reconciler/common/install_test.go:32]: shall we remove this file entirely.

File: pkg/reconciler/common/releases_test.go [hunk 3]
```diff
@@ -48,19 +48,19 @@ func TestListReleases(t *testing.T) {
...
-	util.AssertNotEqual(t, path, expectedPath)
-
-}
+//func TestManifestPath(t *testing.T) {
+//	koPath := "testdata/kodata"
+//	os.Setenv(KoEnvKey, koPath)
+//	defer os.Unsetenv(KoEnvKey)
+//	expectedPath := "testdata/kodata/tekton-trigger/0.15.2"
+//
+//	path := manifestPath(VERSION, &v1alpha1.TektonTrigger{})
```
Comment (nikhil-thomas) [inline pkg/reconciler/common/releases_test.go:57]: why does this patch preserve these lines as comments.

File: pkg/reconciler/kubernetes/tektoninstallerset/install.go [hunk 4]
```diff
@@ -96,35 +97,72 @@ func (i *installer) EnsureDeploymentResources() error {
...
 		}
+	}
 
-		if !isDeploymentAvailable(deployment) {
-			return errors.New("deployment not available")
-		}
+	return nil
+}
+
+func (i *installer) isDeploymentReady(d *unstructured.Unstructured) error {
```
Comment (nikhil-thomas) [inline pkg/reconciler/kubernetes/tektoninstallerset/install.go:151]: nice 🧑‍💻 👍

File: pkg/reconciler/kubernetes/tektontrigger/controller.go [hunk 5]
```diff
@@ -64,24 +62,60 @@ func NewExtendedController(generator common.ExtensionGenerator) injection.Contro
...
+	}
+
+	crd := crds.Resources()[0]
+	version, ok := crd.GetLabels()[releaseLabel]
+	if !ok {
+		return version, fmt.Errorf("failed to find release label on crd")
+	}
+
+	return version, nil
+}
```
Comment (nikhil-thomas) [inline pkg/reconciler/kubernetes/tektontrigger/controller.go:121]: could we move this logic into the common package and reuse with other components.

**Review by vdemeester — COMMENTED** [review]
We do not have a huge test coverage, and this PR removes quite some too, most likely we should port the test to the new struct to make sure we are still covered.

**Comment by sm43** [comment]
> We do not have a huge test coverage, and this PR removes quite some too, most likely we should port the test to the new struct to make sure we are still covered.

@vdemeester yeah test coverage is less. I was going to address it once refactoring is done, but it make sense to start as if there is a bug we will catch.
I will add test cases for new code then we can refactor rest of the code.

**Review by vdemeester — APPROVED** [review]

File: pkg/reconciler/kubernetes/tektontrigger/tektontrigger.go [hunk 6]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+    httt://www.apache.org/licenses/LICENSE-2.0
```
Comment (PuneetPunamiya) [inline pkg/reconciler/kubernetes/tektontrigger/tektontrigger.go:8]: Is this change required ??

File: pkg/reconciler/common/utils.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	crd := crds.Resources()[0]
+	version, ok := crd.GetLabels()[releaseLabel]
+	if !ok {
+		return version, fmt.Errorf("failed to find release label on crd")
+	}
+
+	return version, nil
+}
```
Comment (nikhil-thomas) [inline pkg/reconciler/common/utils.go:43]: this good.

As a future improvement, how about reading the version from the "info" configMap [example](https://github.com/tektoncd/pipeline/blob/main/config/config-info.yaml). 

May be we can set it as unknown, if the configMap is not available. 

What do you think.
if you think that could be a useful improvement, then could you create an issue (#good-first-issue, #help-wanted) to capture it.

else, i am curious to know why. and feel free to resolve this 🙂 👍

Comment (nikhil-thomas) [reply pkg/reconciler/common/utils.go:43]: note: this comment ☝️  is not a blocker.

**Comment by nikhil-thomas** [comment]
LGTM

File: pkg/reconciler/common/utils.go (hunk 7 above)
Comment (sm43) [reply pkg/reconciler/common/utils.go:43]: created https://github.com/tektoncd/operator/issues/413
